### ISSUE TITLE
Integrate token creation with policy assignment

### ISSUE THREAD

**Issue by susanev** [chef/automate#3535]
## overview
As a first pass, the new assign-this-token-to-some policies feature just did the simple thing: fire a request to create a token and fire requests to add it to policies independently. So either could fail and failure cases would not look ideal.

Example: if there is an id conflict on the token, the modal will remain open and report the error, but the policy assignment will proceed and report success "underneath" on the main page. So the policy assignments should wait until the token creation finishes.

But more than that, we want to postpone notification entirely until both the token creation and policy assignments are complete.
Currently, if you create a token and assign it to policies, you can a barrage of notifications, one for each policy assignment. The user should only see 1 notification for the combined actions, one of these:

## design details
- `Created token and assigned to <policy name>.`
- `Created token and assigned to <n> policies>.`
- `Created token but failed to assign to <policy name>.`
- `Created token but failed to assign to <n> policies.`
- `Token creation failed`

**Comment by msorens** [comment]
_Slack conversation 2020-04-30_

msorens  1:41 PM
The problem to solve is this: dispatch 1 request to create token, wait for success, then dispatch n requests for adding policy members, and  after all complete report that n were added (or error, of course).
Three approaches possible:
(a) back-end: a new endpoint that takes a token id and list of policies ids, does the things, and returns yea or nay. Slight drawback here is it does not really fit in our actions scheme and less RESTful: is it a token thing or a policy thing, etc.?
(b) front-end: create a token, if successful, add to policy-1...  if successful, add to policy-n (serial approach).
(c) front-end: create a token, if successful, add to policy-1 through policy-n in parallel, then evaluate the n returned results. The big problem here is  handling the n results, currently all of which would write (really overwrite!) a single ngrx status field.
I am leaning towards (b) as the least amount of work. I do not see a problem doing serially because I do not think n would be very large and adding a member to a policy is very fast.
Thoughts?

susanev  1:43 PM
i also except most folks to be selecting 1 policy for most use cases in the create token flow

idk if this is helpful.. but these are some future things..
- allow folks to edit those policies from the token detail page
- see policies associated with a token on the token list page

blake  1:48 PM
a counter arg to the con of a) is that we already have a number of apis that look like this around membership. Other pros: reduces race condition likelihood, reduces store refresh calls, is more reasonable to build out this api if we’ll have other calls to “it”
more cons: more overhead if we think users will only add one at a time

I am leaning toward A as the “right” thing but B as the expedient thing, so we should think about priority

brenna  1:49 PM
approach B sounds fine. though do we want to be consistent with the way projects and their associated policies are created?

blake  1:50 PM
you mean the auto-created policies?

brenna  1:50 PM
If we were to be consistent with that behavior, we'd basically do approach A but change the existing createToken API instead of creating a new one
1:50
yeah auto-created policies

blake  1:51 PM
these aren’t auto created, though, right? they’re a select of all possible policies? Maybe I need to refresh myself on how we do the projects one

brenna  1:52 PM
I was thinking it's similar in that it's an automated action triggered by the creation of an object

msorens  1:52 PM
Yes, that does make sense @brenna

blake  1:53 PM
but we’d be passing more info, right?
1:53
like the request body would look different, bc we aren’t creating the policies, just the associations

msorens  1:54 PM
Yes.

brenna  1:54 PM
a little more info, but we do the same with a flag on the createProject request
1:54
i'd also be fine changing how the projects pattern works too
1:54
the new API sounds like it might be cleaner

blake  1:56 PM
if B is a quick thing, we can do that to get this over the line. I think this discussion indicates that if we prefer A then that involves some architecting

msorens  1:56 PM
One thing with projects and policies: both are in the same DB so can be transactionalized easily. With tokens and policy membership, those are separate, so no transaction. But I think that is OK to have a successful token creation and failed policy insertions as a return state.

blake  1:56 PM
if that’s true, then it sounds like a separate API to me

msorens  1:58 PM
So a createToken and an addTokenToPolicies?

blake  1:59 PM
But maybe only on the authz side :thinking_face:

msorens  1:59 PM
That would go along with @susanev’s future plans: being able to add/remove tokens from policies.

blake  1:59 PM
I think if we are doing A we should write out the options and discuss them

**Comment by msorens** [comment]
So the only prior art we have is when creating a project the user may opt to also create some associated policies. All of that is done with one API call so I think it makes sense to do the same here, where we are creating a token and associating it with some policies.
On the other hand, if the expectation of most folks selecting just one policy is reasonable, it is not clear to me that any of this extra work is necessary... except for the bug reported in #3537: if you attempt to create a token and get an id conflict, the policy assignment still would happen, which is rather bad. The good thing is that, in isolation, that is an easy fix.

**Comment by susanev** [comment]
this work has been deprioritized.
